###LOAD DATA###

---


Loading the csv data using pandas

In [156]:
import pandas as pd

data = pd.read_csv("/content/customerdemographic.csv")
data

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot        NaN    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3997         3998          Sarene    Woolley       U   
3998         3999       Patrizius        NaN    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases         DOB               job_title  \
0                                      93  1953-10-12     Executive Secretary   
1                                      81  1980-12-16  Administrative Officer   
2                                      61  1954-01-20      Recruiting Manager   
3                                      33  1961-10-03                     NaN   
4                                      56  1977-05-13           Senior Editor   
...                                   ...         ...                     ...   
3995                                    8  1975-08-09   VP Product Management   
3996                                   87  2001-07-13         Statistician II   
3997                                   60         NaN       Assistant Manager   
3998                                   11  1973-10-24                     NaN   
3999                                   76  1991-11-05    Software Engineer IV   

     job_industry_category     wealth_segment deceased_indicator  \
0                   Health      Mass Customer                  N   
1       Financial Services      Mass Customer                  N   
2                 Property      Mass Customer                  N   
3                       IT      Mass Customer                  N   
4                      NaN  Affluent Customer                  N   
...                    ...                ...                ...   
3995                Health      Mass Customer                  N   
3996         Manufacturing     High Net Worth                  N   
3997                    IT     High Net Worth                  N   
3998         Manufacturing  Affluent Customer                  N   
3999                   NaN  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                                 1-Feb      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3995                                          -1.00E+02       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3997                                                NaN       No     NaN  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[4000 rows x 13 columns]

###DATA INFORMATION###

---




In [157]:
print("DATA DESCRIPTION")
print(data.describe())
print()

print("DATATYPES")
print(data.info())
print()

print("COLUMNS")
print(data.columns)
print()

DATA DESCRIPTION
       customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

DATATYPES
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   int64  
 1   first_name                           4000

###HANDLING NULL VALUES###

---



####**Checking for null Values in the dataset**

In [158]:
data.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

####***'last_name'*** Column

Filling the null values in ***last_name*** column with ***'Unknown'*** since the last_name are unique in most cases so imputing it with mode value will not be a good choice and deleting the columns can make are dataset short because their are many null values in last_name column.

In [159]:
data['last_name'] = data['last_name'].fillna("Unkown");
data.isnull().sum()

customer_id                              0
first_name                               0
last_name                                0
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

#### ***'DOB'*** & ***'tenure'*** Column

Dropping the null values in ***DOB*** column because their are only 87 cells in the DOB column with null values and dropping them will not cause any issue as 87 is not a big number of rows in 4000 rows dataset
and by deleting the 87 rows in ***DOB*** column ***tenure*** column null values are also deleted.

In [160]:
# print("Mode value for DOB col is ", data['DOB'].mode())
# data['DOB'] = data['DOB'].fillna("1978-01-30");
data.dropna(subset=['DOB'], inplace=True)
data.isnull().sum()

customer_id                              0
first_name                               0
last_name                                0
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                              497
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                215
owns_car                                 0
tenure                                   0
dtype: int64

#### ***'job_title'*** Column

Filling the null values in ***job_title*** column with mode value that column i.e ***'Business Systems Development Analyst'*** since deleting the columns can make are dataset short because their are total of *497* null values in job_title column.

In [161]:


print("Mode value for job_title col is ", data['job_title'].mode())
print((data['job_title'] == "Business Systems Development Analyst").sum() , "Occurrences");
print()

data['job_title'] = data['job_title'].fillna("Business Systems Development Analyst");
data.isnull().sum()

Mode value for job_title col is  0    Business Systems Development Analyst
1                          Tax Accountant
Name: job_title, dtype: object
43 Occurrences



customer_id                              0
first_name                               0
last_name                                0
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                                0
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                215
owns_car                                 0
tenure                                   0
dtype: int64

##### ***'job_industry_category'*** Column

Filling the null values in ***job_industry_category*** column with mode value that column i.e ***'Manufacturing'*** since deleting the columns can make are dataset short because their are total of *656* null values in job_industry_category column.

In [162]:
print("Mode value for job_industry_category col is ", data['job_industry_category'].mode())
print((data['job_industry_category'] == "Manufacturing").sum() , "Occurrences");
print()

data['job_industry_category'] = data['job_industry_category'].fillna("Manufacturing");
data.isnull().sum()

Mode value for job_industry_category col is  0    Manufacturing
Name: job_industry_category, dtype: object
796 Occurrences



customer_id                              0
first_name                               0
last_name                                0
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                      0
job_title                                0
job_industry_category                    0
wealth_segment                           0
deceased_indicator                       0
default                                215
owns_car                                 0
tenure                                   0
dtype: int64

#### ***'default'*** Column

Filling the null values in ***default*** column with mode value that column i.e ***'1.00E+02'*** since deleting the columns can make are dataset short because their are total of *215* null values in default column.

In [163]:
print("Mode value for default col is ", data['default'].mode())
print((data['default'] == "1.00E+02").sum() , "Occurrences");
print()

data['default'] = data['default'].fillna("1.00E+02");
data.isnull().sum()

Mode value for default col is  0    1.00E+02
Name: default, dtype: object
111 Occurrences



customer_id                            0
first_name                             0
last_name                              0
gender                                 0
past_3_years_bike_related_purchases    0
DOB                                    0
job_title                              0
job_industry_category                  0
wealth_segment                         0
deceased_indicator                     0
default                                0
owns_car                               0
tenure                                 0
dtype: int64

###DATA###


---



Null values is the data are sorted.




In [164]:
data

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp       F   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot     Unkown    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3994         3995           Rusty     Iapico    Male   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3998         3999       Patrizius     Unkown    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases         DOB  \
0                                      93  1953-10-12   
1                                      81  1980-12-16   
2                                      61  1954-01-20   
3                                      33  1961-10-03   
4                                      56  1977-05-13   
...                                   ...         ...   
3994                                   93  1975-12-12   
3995                                    8  1975-08-09   
3996                                   87  2001-07-13   
3998                                   11  1973-10-24   
3999                                   76  1991-11-05   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
3     Business Systems Development Analyst                    IT   
4                            Senior Editor         Manufacturing   
...                                    ...                   ...   
3994                       Staff Scientist         Manufacturing   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   
3998  Business Systems Development Analyst         Manufacturing   
3999                  Software Engineer IV         Manufacturing   

         wealth_segment deceased_indicator  \
0         Mass Customer                  N   
1         Mass Customer                  N   
2         Mass Customer                  N   
3         Mass Customer                  N   
4     Affluent Customer                  N   
...                 ...                ...   
3994      Mass Customer                  N   
3995      Mass Customer                  N   
3996     High Net Worth                  N   
3998  Affluent Customer                  N   
3999  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                                 1-Feb      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3994                                                 á       Yes    14.0  
3995                                          -1.00E+02       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[3913 rows x 13 columns]

###Structuring the ***'gender'*** column

---



In [165]:
gender_structure = {
    'M' : 'Male',
    'F' : 'Female',
    'U' : 'Unknown'
}
data['gender'] = data['gender'].replace(gender_structure)
data

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp  Female   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot     Unkown    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3994         3995           Rusty     Iapico    Male   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3998         3999       Patrizius     Unkown    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases         DOB  \
0                                      93  1953-10-12   
1                                      81  1980-12-16   
2                                      61  1954-01-20   
3                                      33  1961-10-03   
4                                      56  1977-05-13   
...                                   ...         ...   
3994                                   93  1975-12-12   
3995                                    8  1975-08-09   
3996                                   87  2001-07-13   
3998                                   11  1973-10-24   
3999                                   76  1991-11-05   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
3     Business Systems Development Analyst                    IT   
4                            Senior Editor         Manufacturing   
...                                    ...                   ...   
3994                       Staff Scientist         Manufacturing   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   
3998  Business Systems Development Analyst         Manufacturing   
3999                  Software Engineer IV         Manufacturing   

         wealth_segment deceased_indicator  \
0         Mass Customer                  N   
1         Mass Customer                  N   
2         Mass Customer                  N   
3         Mass Customer                  N   
4     Affluent Customer                  N   
...                 ...                ...   
3994      Mass Customer                  N   
3995      Mass Customer                  N   
3996     High Net Worth                  N   
3998  Affluent Customer                  N   
3999  Affluent Customer                  N   

                                                default owns_car  tenure  
0                                                    "'      Yes    11.0  
1                          <script>alert('hi')</script>      Yes    16.0  
2                                                 1-Feb      Yes    15.0  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                   NIL      Yes     8.0  
...                                                 ...      ...     ...  
3994                                                 á       Yes    14.0  
3995                                          -1.00E+02       No    19.0  
3996                                           â¦test⧠     Yes     1.0  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0  
3999                                                0/0       No    11.0  

[3913 rows x 13 columns]

###Derving ***'Age'*** Column from ***'DOB'*** column

---



Changing the format of DOB column to date format using to_datetime function in pandas getting the current year using datetime.now function from panda and subtracting the current year from DOB year to get the age and saving it in a new column name age.

In [166]:
data['DOB']= pd.to_datetime(data['DOB'])
current_year = pd.datetime.now().year
print(current_year);
data['Age'] = current_year - data['DOB'].dt.year
data

2023


<ipython-input-166-24c122359d89>:2: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  current_year = pd.datetime.now().year


customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp  Female   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot     Unkown    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3994         3995           Rusty     Iapico    Male   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3998         3999       Patrizius     Unkown    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases        DOB  \
0                                      93 1953-10-12   
1                                      81 1980-12-16   
2                                      61 1954-01-20   
3                                      33 1961-10-03   
4                                      56 1977-05-13   
...                                   ...        ...   
3994                                   93 1975-12-12   
3995                                    8 1975-08-09   
3996                                   87 2001-07-13   
3998                                   11 1973-10-24   
3999                                   76 1991-11-05   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
3     Business Systems Development Analyst                    IT   
4                            Senior Editor         Manufacturing   
...                                    ...                   ...   
3994                       Staff Scientist         Manufacturing   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   
3998  Business Systems Development Analyst         Manufacturing   
3999                  Software Engineer IV         Manufacturing   

         wealth_segment deceased_indicator  \
0         Mass Customer                  N   
1         Mass Customer                  N   
2         Mass Customer                  N   
3         Mass Customer                  N   
4     Affluent Customer                  N   
...                 ...                ...   
3994      Mass Customer                  N   
3995      Mass Customer                  N   
3996     High Net Worth                  N   
3998  Affluent Customer                  N   
3999  Affluent Customer                  N   

                                                default owns_car  tenure  Age  
0                                                    "'      Yes    11.0   70  
1                          <script>alert('hi')</script>      Yes    16.0   43  
2                                                 1-Feb      Yes    15.0   69  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0   62  
4                                                   NIL      Yes     8.0   46  
...                                                 ...      ...     ...  ...  
3994                                                 á       Yes    14.0   48  
3995                                          -1.00E+02       No    19.0   48  
3996                                           â¦test⧠     Yes     1.0   22  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0   50  
3999                                                0/0       No    11.0   32  

[3913 rows x 14 columns]

####Deleting ***DOB*** column

In [167]:
data.drop(['DOB'] , axis = 1 , inplace= True );
data

customer_id      first_name  last_name  gender  \
0               1         Laraine  Medendorp  Female   
1               2             Eli    Bockman    Male   
2               3           Arlin     Dearle    Male   
3               4          Talbot     Unkown    Male   
4               5  Sheila-kathryn     Calton  Female   
...           ...             ...        ...     ...   
3994         3995           Rusty     Iapico    Male   
3995         3996         Rosalia   Halgarth  Female   
3996         3997          Blanch     Nisuis  Female   
3998         3999       Patrizius     Unkown    Male   
3999         4000           Kippy    Oldland    Male   

      past_3_years_bike_related_purchases  \
0                                      93   
1                                      81   
2                                      61   
3                                      33   
4                                      56   
...                                   ...   
3994                                   93   
3995                                    8   
3996                                   87   
3998                                   11   
3999                                   76   

                                 job_title job_industry_category  \
0                      Executive Secretary                Health   
1                   Administrative Officer    Financial Services   
2                       Recruiting Manager              Property   
3     Business Systems Development Analyst                    IT   
4                            Senior Editor         Manufacturing   
...                                    ...                   ...   
3994                       Staff Scientist         Manufacturing   
3995                 VP Product Management                Health   
3996                       Statistician II         Manufacturing   
3998  Business Systems Development Analyst         Manufacturing   
3999                  Software Engineer IV         Manufacturing   

         wealth_segment deceased_indicator  \
0         Mass Customer                  N   
1         Mass Customer                  N   
2         Mass Customer                  N   
3         Mass Customer                  N   
4     Affluent Customer                  N   
...                 ...                ...   
3994      Mass Customer                  N   
3995      Mass Customer                  N   
3996     High Net Worth                  N   
3998  Affluent Customer                  N   
3999  Affluent Customer                  N   

                                                default owns_car  tenure  Age  
0                                                    "'      Yes    11.0   70  
1                          <script>alert('hi')</script>      Yes    16.0   43  
2                                                 1-Feb      Yes    15.0   69  
3     () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0   62  
4                                                   NIL      Yes     8.0   46  
...                                                 ...      ...     ...  ...  
3994                                                 á       Yes    14.0   48  
3995                                          -1.00E+02       No    19.0   48  
3996                                           â¦test⧠     Yes     1.0   22  
3998                             ¡â¢Â£Â¢â§¶â¢ÂªÂºââ       Yes    10.0   50  
3999                                                0/0       No    11.0   32  

[3913 rows x 13 columns]

###OUTLIERS

---



####outliers found in tenure column at threshold 2

In [168]:
value_column = 'tenure'
outliers = data[(data[value_column] < data[value_column].mean() - 2 * data[value_column].std()) |
              (data[value_column] > data[value_column].mean() + 2 * data[value_column].std())]

print("Outliers:\n")
outliers

Outliers:



customer_id first_name     last_name  gender  \
35             36    Lurette      Stonnell  Female   
38             39   Hunfredo       Smalley    Male   
61             62     Sorcha       Roggers  Female   
259           260     Farand      Marriner  Female   
275           276      Goldi         Dwine  Female   
386           387      Spike        Thieme    Male   
401           402      Berne       Donegan    Male   
483           484      Mason       Kohrsen    Male   
594           595     Nikola        Jeanon    Male   
761           762     Kerwin        Jossel    Male   
763           764     Samuel         Stiff    Male   
783           784   Frederik       Duckett    Male   
814           815      Caryn        Sampey  Female   
970           971      Sarge       Lerohan    Male   
1003         1004      Colas      Pumfrett    Male   
1202         1203   Claresta       Carnier  Female   
1312         1313     Hughie       Laboune    Male   
1365         1366   Zacharia        Rigler    Male   
1442         1443     Angela        Mowles  Female   
1467         1468    Adaline        Hayles  Female   
1492         1493   Lisabeth       Fundell  Female   
1506         1507     Holmes        Benkin    Male   
1520         1521    Pernell       Duffett    Male   
1570         1571     Tadeas         McUre    Male   
1611         1612     Wendel      Headland    Male   
1620         1621     Sibley       Eliyahu  Female   
1727         1728      Pryce        Larner    Male   
1905         1906     Waylan         Katte    Male   
1957         1958     Janina          Keme  Female   
1961         1962     Paddie       Grunder    Male   
2053         2054     Horace     Gilpillan    Male   
2055         2056    Cristin     Smoughton  Female   
2120         2121    Clarita    Jaqueminet  Female   
2131         2132     Davina       Dougary  Female   
2172         2173     Gracia       McAteer  Female   
2399         2400    Lindsay  Stollsteiner    Male   
2416         2417    Cacilia      Swetland  Female   
2566         2567       Tobe        Spirit    Male   
2691         2692    Jacklyn      Bromehed  Female   
2766         2767      Sybil      Scramage  Female   
2810         2811     Sharia       Allkins  Female   
2893         2894    Gilburt        Unkown    Male   
2999         3000       Tann        McCard    Male   
3167         3168      Johan    Dudderidge    Male   
3217         3218      Conny      Starbeck  Female   
3325         3326        Wes        Crotch    Male   
3393         3394      Barde       Spoward    Male   
3403         3404  Thomasine       McCloch  Female   
3628         3629       Vale       Starcks  Female   
3742         3743       Riki       Aimable  Female   
3785         3786     Barnie    Blenkensop    Male   
3911         3912       Ryon      Darridon    Male   
3913         3914     Izabel       Roelofs  Female   
3929         3930    Blancha         Baldi  Female   
3967         3968  Alexandra         Kroch  Female   

      past_3_years_bike_related_purchases  \
35                                     33   
38                                      5   
61                                     38   
259                                    47   
275                                    47   
386                                    34   
401                                     9   
483                                    64   
594                                    72   
761                                    91   
763                                    40   
783                                    10   
814                                    47   
970                                    11   
1003                                   35   
1202                                   38   
1312                                   59   
1365                                   10   
1442                                   47   
1467                                   37   
1492                                

####outliers found in age column at threshold 3

In [169]:
value_column = 'Age'
outliers = data[(data[value_column] < data[value_column].mean() - 2 * data[value_column].std()) |
              (data[value_column] > data[value_column].mean() + 2 * data[value_column].std())]

print("Outliers:\n")
outliers

Outliers:



customer_id first_name last_name   gender  \
33             34   Jephthah  Bachmann  Unknown   
657           658       Donn   Bonnell     Male   
719           720     Darrel     Canet     Male   
1091         1092     Katlin   Creddon   Female   
2412         2413      Abbey    Murrow     Male   
3409         3410    Merrili   Brittin   Female   

      past_3_years_bike_related_purchases  \
33                                     59   
657                                    38   
719                                    67   
1091                                   56   
2412                                   27   
3409                                   93   

                                 job_title job_industry_category  \
33                         Legal Assistant                    IT   
657                         Tax Accountant         Manufacturing   
719                     Recruiting Manager                Retail   
1091                    VP Quality Control                Retail   
2412              Environmental Specialist         Manufacturing   
3409  Business Systems Development Analyst              Property   

         wealth_segment deceased_indicator             default owns_car  \
33    Affluent Customer                  N            1.00E+02       No   
657   Affluent Customer                  N  ì¬íê³¼íì ì´íì°êµ¬ì      Yes   
719   Affluent Customer                  N  1;DROP TABLE users       No   
1091      Mass Customer                  N  ì¬íê³¼íì ì´íì°êµ¬ì       No   
2412     High Net Worth                  N                  á       Yes   
3409      Mass Customer                  N                   á       No   

      tenure  Age  
33      20.0  180  
657      8.0   79  
719      6.0   92  
1091     5.0   88  
2412    17.0   80  
3409    16.0   83

###SAVING THE CLEANED DATA

---



In [179]:
data.to_csv(r"cleaned_data1.csv");